### PR TITLE
Add internal/worker/worker test

### PR DESCRIPTION
**Pull request by kevindiu**
Author review required.

### Description:

This PR includes the following changes:
- implements internal/worker test code
- create test/comparator package and implement go-cmp comparator
- missing `close(ech)` in `startJobLoop()`
 

### Environment:

- Go Version: 1.14.4
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.0

### Types of changes:

- [x] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/test/comparator/standard.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package comparator
+
+import (
+	"sync/atomic"
+
+	"github.com/vdaas/vald/internal/errgroup"
+)
```
Comment (hlts2) [inline internal/test/comparator/standard.go:7]: Could you please add alias of `go-comp`.

Comment (kevindiu) [reply internal/test/comparator/standard.go:7]: Sorry i am not sure what you mean 🤔

Comment (hlts2) [reply internal/test/comparator/standard.go:7]: for example.
```
type Option = cmp.Option
```

Comment (kevindiu) [reply internal/test/comparator/standard.go:8]: fixed.
https://github.com/vdaas/vald/pull/602/commits/4d7920ab0518e0b39e0652da7213f0266f33922d

File: internal/worker/worker.go [hunk 2]
```diff
@@ -78,6 +81,7 @@ func New(opts ...WorkerOption) (Worker, error) {
 	return w, nil
 }
 
+// Start start execute jobs in the worker queue. It returns the error channel that the job return, and the error if start failed.
```
Comment (vankichi) [inline internal/worker/worker.go:84]:
```suggestion
// Start starts execute jobs in the worker queue. It returns the error channel that the job return, and the error if start failed.
```

File: internal/worker/worker.go [hunk 3]
```diff
@@ -54,6 +56,7 @@ type worker struct {
 	completedCount uint64
 }
 
+// New initialize and return the worker, or return initialization error if occurred.
```
Comment (vankichi) [inline internal/worker/worker.go:59]:
```suggestion
// New initializes and return the worker, or return initialization error if occurred.
```

File: internal/worker/worker.go [hunk 4]
```diff
@@ -29,8 +29,10 @@ import (
 	"github.com/vdaas/vald/internal/safety"
 )
 
+// JobFunc represent the function of a job that work in the worker.
 type JobFunc func(context.Context) error
 
+// Worker represent the worker interface to execute jobs.
```
Comment (vankichi) [inline internal/worker/worker.go:35]:
```suggestion
// Worker represents the worker interface to execute jobs.
```

File: internal/worker/worker.go [hunk 5]
```diff
@@ -29,8 +29,10 @@ import (
 	"github.com/vdaas/vald/internal/safety"
 )
 
+// JobFunc represent the function of a job that work in the worker.
```
Comment (vankichi) [inline internal/worker/worker.go:32]:
```suggestion
// JobFunc represents the function of a job that works in the worker.
```

File: internal/worker/worker.go [hunk 6]
```diff
@@ -176,34 +184,44 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
 	return ech
 }
 
+// Pause stop allowing new job to be dispatched to the worker.
```
Comment (vankichi) [inline internal/worker/worker.go:187]:
```suggestion
// Pause stops allowing new job to be dispatched to the worker.
```

File: internal/worker/worker.go [hunk 7]
```diff
@@ -176,34 +184,44 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 func (w *worker) Pause() {
 	w.running.Store(false)
 }
 
+// Resume resume to allow new job to be dispatched to the worker.
```
Comment (vankichi) [inline internal/worker/worker.go:192]:
```suggestion
// Resume resumes to allow new jobs to be dispatched to the worker.
```

File: internal/worker/worker.go [hunk 8]
```diff
@@ -176,34 +184,44 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 func (w *worker) Resume() {
 	w.running.Store(true)
 }
 
+// IsRunning return if the worker is running or not.
```
Comment (vankichi) [inline internal/worker/worker.go:197]:
```suggestion
// IsRunning returns if the worker is running or not.
```

File: internal/worker/worker.go [hunk 9]
```diff
@@ -176,34 +184,44 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 func (w *worker) IsRunning() bool {
 	return w.running.Load().(bool)
 }
 
+// Name return the worker name.
```
Comment (vankichi) [inline internal/worker/worker.go:202]:
```suggestion
// Name returns the worker name.
```

File: internal/worker/worker.go [hunk 10]
```diff
@@ -176,34 +184,44 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 func (w *worker) Name() string {
 	return w.name
 }
 
+// Len return the length of the worker queue.
```
Comment (vankichi) [inline internal/worker/worker.go:207]:
```suggestion
// Len returns the length of the worker queue.
```

File: internal/worker/worker.go [hunk 11]
```diff
@@ -176,34 +184,44 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 func (w *worker) Len() uint64 {
 	return w.queue.Len()
 }
 
+// TotalRequested return the number of job that dispatched to the worker.
```
Comment (vankichi) [inline internal/worker/worker.go:212]:
```suggestion
// TotalRequested returns the number of jobs that dispatched to the worker.
```

File: internal/worker/worker.go [hunk 12]
```diff
@@ -176,34 +184,44 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 func (w *worker) TotalRequested() uint64 {
 	return atomic.LoadUint64(&w.requestedCount)
 }
 
+// TotalCompleted return the number of completed job.
```
Comment (vankichi) [inline internal/worker/worker.go:217]:
```suggestion
// TotalCompleted returns the number of completed jobs.
```

File: internal/worker/worker.go [hunk 13]
```diff
@@ -176,34 +184,44 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 func (w *worker) TotalCompleted() uint64 {
 	return atomic.LoadUint64(&w.completedCount)
 }
 
+// Dispatch dispatch the job to the worker and waiting for worker to process it.
```
Comment (vankichi) [inline internal/worker/worker.go:222]:
```suggestion
// Dispatch dispatches the job to the worker and waiting for the worker to process it.
```

File: internal/worker/worker.go [hunk 14]
```diff
@@ -176,34 +184,44 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
+// The job error is push to the error channel that Start() return.
```
Comment (vankichi) [inline internal/worker/worker.go:223]:
```suggestion
// The job error is pushed to the error channel that Start() return.
```

File: internal/worker/worker.go [hunk 15]
```diff
@@ -176,34 +184,44 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
+// This function will return error if the job cannot be dispatch to the worker queue, or the worker is not running.
```
Comment (vankichi) [inline internal/worker/worker.go:224]:
```suggestion
// This function will return an error if the job cannot be dispatch to the worker queue, or the worker is not running.
```

File: internal/worker/worker_test.go [hunk 16]
```diff
@@ -49,42 +58,111 @@ func TestNew(t *testing.T) {
...
+
+		egComparator := func(x, y errgroup.Group) bool {
+			return reflect.DeepEqual(x, y)
+		}
+		atomicValueComparator := func(x, y atomic.Value) bool {
+			return reflect.DeepEqual(x.Load(), y.Load())
+		}
+		want := w.want.(*worker)
+
+		queueOpts := []cmp.Option{
```
Comment (hlts2) [inline internal/worker/worker_test.go:70]: could you please use Option in `internal/test/comparator/standard.go`

File: internal/worker/worker_test.go [hunk 17]
```diff
@@ -49,42 +58,111 @@ func TestNew(t *testing.T) {
...
+			cmp.AllowUnexported(*(want.queue.(*queue))),
```
Comment (hlts2) [inline internal/worker/worker_test.go:71]: could you please use AllowUnexported in internal/test/comparator/standard.go

File: internal/worker/worker_test.go [hunk 18]
```diff
@@ -22,13 +22,22 @@ import (
 	"reflect"
 	"sync/atomic"
 	"testing"
+	"time"
 
+	"github.com/google/go-cmp/cmp"
```
Comment (hlts2) [inline internal/worker/worker_test.go:27]:
```suggestion
	"github.cominternal/test/comparator"
```

File: internal/worker/worker_test.go [hunk 19]
```diff
@@ -247,62 +367,216 @@ func Test_worker_startJobLoop(t *testing.T) {
 		afterFunc  func(args)
 	}
 	defaultCheckFunc := func(w want, got <-chan error) error {
-		if !reflect.DeepEqual(got, w.want) {
-			return errors.Errorf("got = %v, want %v", got, w.want)
+		ecComparator := func(x, y <-chan error) bool {
```
Comment (hlts2) [inline internal/worker/worker_test.go:370]: x and y are too abstract. I will have to go back to the caller, so rename it appropriately.

File: internal/worker/worker.go [hunk 20]
```diff
@@ -126,17 +131,22 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 			select {
 			case <-ctx.Done():
+				var egErr error
 				if err = ctx.Err(); err != nil {
-					return errors.Wrap(eg.Wait(), err.Error())
+					egErr = eg.Wait()
+					return errors.Wrap(egErr, err.Error())
 				}
-				return eg.Wait()
+				return egErr
```
Comment (hlts2) [inline internal/worker/worker.go:149]: I think that even when ctx.Err() is nil, you use eg.Wait()

File: internal/worker/worker.go [hunk 21]
```diff
@@ -152,12 +162,11 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
 
 			if job != nil {
 				eg.Go(safety.RecoverFunc(func() (err error) {
-					defer atomic.AddUint64(&w.completedCount, 1)
-					err = job(ctx)
-					if err != nil {
+					if err = job(ctx); err != nil {
 						log.Debugf("an error occurred while executing a job: %s", err)
 						ech <- err
```
Comment (hlts2) [inline internal/worker/worker.go:167]: you think you use `select` with `ctx.Done` channel

File: internal/worker/worker.go (hunk 20 above)
Comment (kpango) [reply internal/worker/worker.go:149]: why you delete Wait?????

Comment (kpango) [reply internal/worker/worker.go:149]: please do not change goroutine shutdown logic with no proposals

File: internal/worker/worker.go [hunk 22]
```diff
@@ -152,12 +162,11 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 					}
+					atomic.AddUint64(&w.completedCount, 1)
```
Comment (kpango) [inline internal/worker/worker.go:169]: I think defer is better if ech was closed, it makes runtime panic so counter is not increase even jos finished

File: internal/worker/worker.go (hunk 20 above)
Comment (kevindiu) [reply internal/worker/worker.go:149]: will fix it now!

File: internal/worker/worker.go (hunk 22 above)
Comment (kpango) [inline internal/worker/worker.go:169]:
```suggestion
					defer atomic.AddUint64(&w.completedCount, 1)
					if err = job(ctx); err != nil {
						log.Debugf("an error occurred while executing a job: %s", err)
						ech <- err
					}
```

File: internal/worker/worker.go [hunk 23]
```diff
@@ -126,17 +131,22 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 		eg, ctx := errgroup.New(ctx)
 		eg.Limitation(w.limitation)
+
 		limitation := make(chan struct{}, w.limitation)
 		defer close(limitation)
+
 		for {
 			select {
 			case <-ctx.Done():
+				var egErr error
```
Comment (kpango) [inline internal/worker/worker.go:144]:
```suggestion
				egErr := eg.Wait()
```

File: internal/worker/worker.go [hunk 24]
```diff
@@ -126,17 +131,22 @@ func (w *worker) startJobLoop(ctx context.Context) <-chan error {
...
 				if err = ctx.Err(); err != nil {
-					return errors.Wrap(eg.Wait(), err.Error())
+					egErr = eg.Wait()
+					return errors.Wrap(egErr, err.Error())
```
Comment (kpango) [inline internal/worker/worker.go:147]:
```suggestion
					return errors.Wrap(egErr, err.Error())
```

File: internal/worker/worker_test.go (hunk 19 above)
Comment (kevindiu) [reply internal/worker/worker_test.go:371]: do you have any idea what name should we use?
I think this func is small enough and should be easy to understand...

Comment (hlts2) [reply internal/worker/worker_test.go:371]: I recommend that `x is want`, `y is got`

Comment (kevindiu) [reply internal/worker/worker_test.go:371]: actually I followed the comparator implementation that it don't care the order of the parameters.
I will change to not using the comparator implementation

Comment (kevindiu) [reply internal/worker/worker_test.go]: fixed

File: internal/worker/worker_test.go [hunk 25]
```diff
@@ -49,42 +59,111 @@ func TestNew(t *testing.T) {
...
+			want: want{
+				want: &worker{
+					name:       "worker",
+					limitation: 10,
+					eg:         errgroup.Get(),
+					running: func() atomic.Value {
+						v := new(atomic.Value)
+						v.Store(false)
+						return *v
+					}(),
```
Comment (vankichi) [inline internal/worker/worker_test.go:104]:
```suggestion
					running: func() (v atomic.Value) {
						v.Store(false)
						return v
					}(),
```
@hlts2 commented to me #606

Comment (kevindiu) [reply internal/worker/worker_test.go]: I think both are the same. anyway I fixed it.

Comment (vankichi) [reply internal/worker/worker_test.go]: yes. I think so.

**Comment by rinx** [comment]
@kevindiu Could you add lines to the ignore paths in GO_SOURCES and GO_OPTION_SOURCES variables here?
https://github.com/vdaas/vald/blob/031a221446410e684d6a27b00b1c36d6913da670/Makefile#L113
https://github.com/vdaas/vald/blob/031a221446410e684d6a27b00b1c36d6913da670/Makefile#L132

```go
		-not -path './internal/core/ngt/*' \
		-not -path './internal/test/comparator/*' \
```

**Comment by kevindiu** [comment]
> @kevindiu Could you add lines to the ignore paths in GO_SOURCES and GO_OPTION_SOURCES variables here?
> https://github.com/vdaas/vald/blob/031a221446410e684d6a27b00b1c36d6913da670/Makefile#L113
> 
> 
> https://github.com/vdaas/vald/blob/031a221446410e684d6a27b00b1c36d6913da670/Makefile#L132
> 
> ```go
> 		-not -path './internal/core/ngt/*' \
> 		-not -path './internal/test/comparator/*' \
> ```

Thank you for your comment. I think ignoring `./internal/test/*` is better, what do you think?

**Comment by kevindiu** [comment]
> @kevindiu Could you add lines to the ignore paths in GO_SOURCES and GO_OPTION_SOURCES variables here?
> https://github.com/vdaas/vald/blob/031a221446410e684d6a27b00b1c36d6913da670/Makefile#L113
> 
> 
> https://github.com/vdaas/vald/blob/031a221446410e684d6a27b00b1c36d6913da670/Makefile#L132
> 
> ```go
> 		-not -path './internal/core/ngt/*' \
> 		-not -path './internal/test/comparator/*' \
> ```

@rinx 
Thank you for your comment. I think ignoring `./internal/test/*` is better, what do you think?

**Comment by rinx** [comment]
@kevindiu I don't know. Either is ok.

**Comment by rinx** [comment]
Thanks!

File: internal/worker/queue_mock_test.go [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package worker
+
+import "context"
+
+var (
+	DefaultStartFunc = func(context.Context) (<-chan error, error) {
```
Comment (hlts2) [inline internal/worker/queue_mock_test.go:21]: Why do you need this ? (Default〇〇Func)

Comment (kevindiu) [reply internal/worker/queue_mock_test.go:21]: I can use it to initialize `QueueMock`

**Pull request merged**